tap_bing_hotel_ads/__init__.py

```diff
@@ -0,0 +1,196 @@
+from zipfile import ZipFile
+import arrow
+
+import singer
+from singer import metrics
+from singer import bookmarks
+from singer import utils
+from singer import metadata
+
+import tap_bing_hotel_ads.reports
+
+
+LOGGER = singer.get_logger()
+
+REQUIRED_CONFIG_KEYS = [
+    "start_date",
+    "customer_id",
+    "account_id",
+    "oauth_access_token",
+    "oauth_refresh_token",
+    "ms_client_id",
+]
+
+DATE_FORMAT='%Y-%m-%d'
+SESSION = requests.session()
+CONFIG = {}
+DEFAULT_COLS = ["HotelId", "Clicks"] 
+
+# ~10 min polling timeout
+MAX_NUM_REPORT_POLLS = 120
+REPORT_POLL_SLEEP = 5
+
+def get_abs_path(path):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
+
+def load_schema(entity):
+    return utils.load_json(get_abs_path("schemas/{}.json".format(entity)))
+
+DEFAULT_USER_AGENT = 'Singer.io Bing Hotel Ads Tap'
+
+def get_oauth_client():
+    token = {
+        'access_token': CONFIG['oauth_access_token'],
+        'refresh_token': CONFIG['oauth_refresh_token'],
+        'token_type': 'Bearer',
+        'expires_in': '0',
+    }
+    refresh_url = 'https://login.microsoftonline.com/common/oauth2/v2.0/token'
+    extra={'client_id': CONFIG['ms_client_id']}
+    client_id = CONFIG['ms_client_id']
+    def token_saver(new_token):
+        token = new_token
+        CONFIG['oauth_access_token'] = token['access_token']
+        CONFIG['oauth_refresh_token'] = token['refresh_token']
+
+    session = OAuth2Session(client_id, token=token, auto_refresh_url=refresh_url, auto_refresh_kwargs=extra, token_updater=token_saver)
+    return session
+
+async def poll_report(customer_id, account_id, job_id, start_date, end_date):
+    download_url = None
+    poll_url = "https://partner.api.bingads.microsoft.com/Travel/v1/Customers({})/Accounts({})/ReportJobs('{}')".format(customer_id, account_id, job_id)
+    with metrics.job_timer('generate_report'):
+        for i in range(1, MAX_NUM_REPORT_POLLS + 1):
+            LOGGER.info('Polling report job {}/{} - {} - from {} to {}'.format(
+                i,
+                MAX_NUM_REPORT_POLLS,
+                job_id,
+                start_date,
+                end_date))
+            response = SESSION.get(poll_url)
+            resp_obj = response.json()
+            if resp_obj["Status"] == 'Completed':
+                if resp_obj["Url"]:
+                    download_url = resp_obj["Url"]
+                else:
+                    LOGGER.info('No results for report: {} - from {} to {}'.format(
+                        job_id,
+                        start_date,
+                        end_date))
+                break
+
+            if i == MAX_NUM_REPORT_POLLS:
+                LOGGER.info('Generating report timed out: {} - from {} to {}'.format(
+                        job_id,
+                        start_date,
+                        end_date))
+                return False, ''
+            else:
+                await asyncio.sleep(REPORT_POLL_SLEEP)
```

Comment:
Did you test this? I think you should be using an async with 

> async with metrics.job_timer('generate_report'):

Reply:
I think this is fine. It looks like async with just runs the resource acquisition in a different context?

https://docs.python.org/3/reference/compound_stmts.html#the-async-with-statement